sample/weighted.go

```diff
@@ -0,0 +1,118 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import (
+	"math/rand"
+
+	"github.com/gonum/floats"
+)
+
+// Weighted provides sampling without replacement from a collection of items with
+// non-uniform probability.
+type Weighted struct {
+	weights []float64
+	// heap is a weight heap. See comments
```

Comment:
1) I'm not sure what you mean by this.

2) Is actually not correct; it does keep a sorted list of weights - admittedly not the weights given to the type, but the totaled weights.

3) I guess this is true, but I think it is generally unerstood that a heap is really just a linear representation of a tree.

4) That is because the container/heap mutate by deleting elements and needs to reattain the heap invariants by reordering, here we mutate weight and reattain invariants by adjusting.

Notwithstanding all these, I will expand the explanation.